### ISSUE TITLE
If dispatchOptionsRequest is disabled, will CORS settings be effective?

### ISSUE THREAD

**Issue by laszlocsontos** [laszlocsontos/relinkr#84]
Discovered during #80 that controllers get executed for pre-flight requests.

```
2019-09-02 07:26:04 default[20190901t135225]  "OPTIONS /v1/links?page=0&size=10&bust=1567409164903" 200
2019-09-02 07:26:05 default[20190901t135225]  2019-09-02 07:26:05.422 ERROR 1 --- [nio-8080-exec-5] io.relinkr.core.web.RestErrorHandler     : null
2019-09-02 07:26:05 default[20190901t135225]
2019-09-02 07:26:05 default[20190901t135225]  java.lang.NullPointerException: null
  	at io.relinkr.link.web.LinkResource.of(LinkResource.java:114) ~[classes!/:1.0.0-SNAPSHOT]
  	at io.relinkr.link.web.LinkResourceAssembler.instantiateResource(LinkResourceAssembler.java:68) ~[classes!/:1.0.0-SNAPSHOT]
  	at io.relinkr.link.web.LinkResourceAssembler.instantiateResource(LinkResourceAssembler.java:38) ~[classes!/:1.0.0-SNAPSHOT]
  	at org.springframework.hateoas.mvc.IdentifiableResourceAssemblerSupport.createResourceWithId(IdentifiableResourceAssemblerSupport.java:72) ~[spring-hateoas-0.25.1.RELEASE.jar!/:na]
  	at org.springframework.hateoas.mvc.IdentifiableResourceAssemblerSupport.createResource(IdentifiableResourceAssemblerSupport.java:63) ~[spring-hateoas-0.25.1.RELEASE.jar!/:na]
  	at org.springframework.hateoas.mvc.IdentifiableResourceAssemblerSupport.createResource(IdentifiableResourceAssemblerSupport.java:59) ~[spring-hateoas-0.25.1.RELEASE.jar!/:na]
  	at io.relinkr.link.web.LinkResourceAssembler.toResource(LinkResourceAssembler.java:55) ~[classes!/:1.0.0-SNAPSHOT]
  	at io.relinkr.link.web.LinkResourceAssembler.toResource(LinkResourceAssembler.java:38) ~[classes!/:1.0.0-SNAPSHOT]
  	at org.springframework.data.web.PagedResourcesAssembler.createResource(PagedResourcesAssembler.java:208) ~[spring-data-commons-2.1.6.RELEASE.jar!/:2.1.6.RELEASE]
  	at org.springframework.data.web.PagedResourcesAssembler.toResource(PagedResourcesAssembler.java:120) ~[spring-data-commons-2.1.6.RELEASE.jar!/:2.1.6.RELEASE]
  	at io.relinkr.link.web.LinkResourceController.fetchLinks(LinkResourceController.java:163)
```

The default value of `dispatchOptionsRequest` has been `true` since Spring 4.3. That's why `LinkResourceController` get executed for pre-flight requests. When disabled, will [CORS settings](https://github.com/laszlocsontos/relinkr/blob/3523a2ffa3b3ff1b2974a754e4fdbd541765205f/relinkr-api/src/main/java/io/relinkr/core/web/WebMvcConfig.java#L85) still be effective?

**Comment by djavorszky** [comment]
Hi @laszlocsontos,

I tested this and it seems that CORS settings are still in place. That said, please validate what I've done as I'm not sure it was the proper way :)

First, without changing anything, I tested the CORS responses by manually sending a CORS request to /v1/links/1 (see https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request)

```
curl -v --insecure -X OPTIONS \
  https://localhost:8443/v1/links/1 \
  -H 'Access-Control-Request-Headers: x-requested-with,origin' \
  -H 'Access-Control-Request-Method: DELETE' \
  -H 'Host: localhost:8443' \
  -H 'Origin: https://localhost:9443'
```

Successful response: 
```
< HTTP/1.1 200
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Access-Control-Allow-Origin: https://localhost:9443
< Access-Control-Allow-Methods: GET,POST,PATCH,PUT,DELETE
< Access-Control-Allow-Headers: x-requested-with, origin
< Access-Control-Allow-Credentials: true
< Access-Control-Max-Age: 1800
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< Strict-Transport-Security: max-age=31536000 ; includeSubDomains
< X-Frame-Options: DENY
< Content-Length: 0
< Date: Sat, 14 Sep 2019 08:14:02 GMT
```

I then modified the request so that `Origin: https://some-other.com` and saw that CORS was in effect:

```
< HTTP/1.1 403
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< Strict-Transport-Security: max-age=31536000 ; includeSubDomains
< X-Frame-Options: DENY
< Content-Length: 20
< Date: Sat, 14 Sep 2019 08:16:51 GMT
```

I've then added the following code snippet to `WebMvcConfig` class for testing purposes:

```
  @Bean
  public DispatcherServlet dispatcherServlet() {
    DispatcherServlet dispatcherServlet = new DispatcherServlet();

    dispatcherServlet.setDispatchOptionsRequest(false);

    return dispatcherServlet;
  }
```

After re-running my test above, I got the very same results. I'm not sure if the `dispatchOptionsRequest` was in effect though (I verified via debugging that the method does get called), nor whether this is the proper way to do it :-)

Hope this helps!

**Comment by laszlocsontos** [comment]
Nice work @djavorszky!

You don't have to expose `DispatcherServlet` as a bean, because in case of Spring Boot there are quite a few auto-configurations in place. Specifically, [`DispatcherServletAutoConfiguration`](https://github.com/spring-projects/spring-boot/blob/v2.1.4.RELEASE/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/DispatcherServletAutoConfiguration.java#L101) is responsible for doing just that.

All you need to do is to set `spring.mvc.dispatch-options-request` to false in `application.yml`.

**Comment by djavorszky** [comment]
Thanks! I went another way for testing this time:

1. Change the default value for `relinkr.frontend.base-url` in `application.yml` to something else
2. Load the UI and see that the request fail as expected
3. Change `spring.mvc.dispatch-options-request` to `false` in `application.yml`

After restarting, all requests were still denied from localhost:9443, as expected.

I could not reproduce the NPE though - is that the original issue from #80 that was fixed by #83?

That said, the OPTIONS requests are still coming from the frontend. What's the expected outcome here? :)

**Comment by laszlocsontos** [comment]
If you change `relinkr.frontend.base-url` to something else, requests from the UI will fail regardless of the value of `spring.mvc.dispatch-options-request`.

**Comment by djavorszky** [comment]
Right, makes sense. During my testing, I've seen that regardless of what value `sprng.mvc.dispatch-options-request` holds, the following are two:
1. CORS settings are in place
2. The controller does not get called.

Looking at the code, the controllers shouldn't be called either, as they have not specified `OPTIONS` as a http method that they should handle.

As a last test, I added a breakpoint to `LinkResourcesController#fetchLinks`, then navigated to the links page (with the `relinkr.frontend.base-url` set correctly with regards to CORS requests).

Once the breakpoint was caught, I checked the call stack and saw that regardless of the value of `spring.mvc.dispatch-options-request`, the controller was only called for the `GET` request.
